From b0492218b02d039db8380cc86e6a4528f455a248 Mon Sep 17 00:00:00 2001
From: Uri Mashiach <uri.mashiach@compulab.co.il>
Date: Mon, 1 Jul 2019 13:52:19 +0300
Subject: [PATCH 2/2] sb-imx8x: add basic support

Basic support for the SB-iMX8X board.
The board is used as the evaluation baseboard for the CL-SOM-iMX8X module.

Add the following support:
* MIPI DSI to DHMI bridge
* 2 CAN bus interfaces
* SD card

Signed-off-by: Uri Mashiach <uri.mashiach@compulab.co.il>
---
 arch/arm64/boot/dts/compulab/Makefile     |   3 +-
 arch/arm64/boot/dts/compulab/sb-imx8x.dts | 253 ++++++++++++++++++++++++++++++
 arch/arm64/configs/cl-som-imx8x_defconfig |   5 +-
 3 files changed, 259 insertions(+), 2 deletions(-)
 create mode 100644 arch/arm64/boot/dts/compulab/sb-imx8x.dts

diff --git a/arch/arm64/boot/dts/compulab/Makefile b/arch/arm64/boot/dts/compulab/Makefile
index 0f7a5947e795..040a0730be9f 100644
--- a/arch/arm64/boot/dts/compulab/Makefile
+++ b/arch/arm64/boot/dts/compulab/Makefile
@@ -1,4 +1,5 @@
-dtb-$(CONFIG_ARCH_FSL_IMX8QXP) += cl-som-imx8x.dtb
+dtb-$(CONFIG_ARCH_FSL_IMX8QXP) += cl-som-imx8x.dtb \
+				  sb-imx8x.dtb
 
 always         := $(dtb-y)
 subdir-y       := $(dts-dirs)
diff --git a/arch/arm64/boot/dts/compulab/sb-imx8x.dts b/arch/arm64/boot/dts/compulab/sb-imx8x.dts
new file mode 100644
index 000000000000..bd1209056055
--- /dev/null
+++ b/arch/arm64/boot/dts/compulab/sb-imx8x.dts
@@ -0,0 +1,253 @@
+/*
+ * Copyright 2019 CompuLab
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ */
+
+#include "cl-som-imx8x.dts"
+
+/ {
+	model = "CompuLab CL-SOM-iMX8X on SB-iMX8X";
+	compatible = "compulab,sbc-imx8x", "compulab,cl-som-imx8x", "fsl,imx8qxp";
+};
+
+&usdhc2 {
+	pinctrl-names = "default", "state_100mhz", "state_200mhz";
+	pinctrl-0 = <&pinctrl_usdhc2>, <&pinctrl_usdhc2_gpio>;
+	pinctrl-1 = <&pinctrl_usdhc2_100mhz>, <&pinctrl_usdhc2_gpio>;
+	pinctrl-2 = <&pinctrl_usdhc2_200mhz>, <&pinctrl_usdhc2_gpio>;
+	bus-width = <4>;
+	cd-gpios = <&gpio4 22 GPIO_ACTIVE_LOW>;
+	wp-gpios = <&gpio4 21 GPIO_ACTIVE_HIGH>;
+	status = "okay";
+};
+
+&i2c1 {
+	eeprom@54 {
+		compatible = "atmel,24c08";
+		reg = <0x54>;
+		pagesize = <16>;
+	};
+
+	pca9555_sb: pca9555_sb@20 {
+		compatible = "nxp,pca9555";
+		gpio-controller;
+		#gpio-cells = <2>;
+		reg = <0x20>;
+	};
+};
+
+/* GPU */
+&gpu_3d0 {
+        status = "okay";
+};
+
+&imx8_gpu_ss {
+        status = "okay";
+};
+
+&pixel_combiner {
+        status = "okay";
+};
+
+&prg1 {
+        status = "okay";
+};
+
+&prg2 {
+        status = "okay";
+};
+
+&prg3 {
+        status = "okay";
+};
+
+&prg4 {
+        status = "okay";
+};
+
+&prg5 {
+        status = "okay";
+};
+
+&prg6 {
+        status = "okay";
+};
+
+&prg7 {
+        status = "okay";
+};
+
+&prg8 {
+        status = "okay";
+};
+
+&prg9 {
+        status = "okay";
+};
+
+&dpr1_channel1 {
+	status = "okay";
+};
+
+&dpr1_channel2 {
+	status = "okay";
+};
+
+&dpr1_channel3 {
+	status = "okay";
+};
+
+&dpr2_channel1 {
+	status = "okay";
+};
+
+&dpr2_channel2 {
+	status = "okay";
+};
+
+&dpr2_channel3 {
+	status = "okay";
+};
+
+/* Display */
+&dpu1 {
+	status = "okay";
+};
+
+/* DSI/LVDS port 0 */
+&i2c0_mipi_lvds0 {
+	#address-cells = <1>;
+	#size-cells = <0>;
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_i2c0_mipi_lvds0>;
+	clock-frequency = <100000>;
+	status = "okay";
+
+	adv_bridge0: adv7535@3d {
+		compatible = "adi,adv7535", "adi,adv7533";
+		reg = <0x3d>;
+		adi,dsi-lanes = <4>;
+		adi,dsi-channel = <1>;
+		interrupt-parent = <&gpio1>;
+		interrupts = <28 IRQ_TYPE_LEVEL_LOW>;
+		status = "okay";
+
+		port {
+			adv7535_1_in: endpoint {
+				remote-endpoint = <&mipi_dsi_bridge1_out>;
+			};
+		};
+	};
+
+};
+
+&mipi_dsi1 {
+	pwr-delay = <10>;
+	status = "okay";
+};
+
+&mipi_dsi_bridge1 {
+	status = "okay";
+
+	port@1 {
+		mipi_dsi_bridge1_out: endpoint {
+			remote-endpoint = <&adv7535_1_in>;
+		};
+	};
+};
+
+&mipi_dsi_phy1 {
+	status = "okay";
+};
+
+/* CAN bus */
+&flexcan1 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan1>;
+	status = "okay";
+};
+
+&flexcan3 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&pinctrl_flexcan3>;
+	status = "okay";
+};
+
+&iomuxc {
+	pinctrl_usdhc2_gpio: usdhc2gpiogrp {
+		fsl,pins = <
+			SC_P_USDHC1_RESET_B_LSIO_GPIO4_IO19	0x00000021
+			SC_P_USDHC1_WP_LSIO_GPIO4_IO21		0x00000021
+			SC_P_USDHC1_CD_B_LSIO_GPIO4_IO22	0x00000021
+		>;
+	};
+
+	pinctrl_usdhc2: usdhc2grp {
+		fsl,pins = <
+			SC_P_USDHC1_CLK_CONN_USDHC1_CLK		0x06000041
+			SC_P_USDHC1_CMD_CONN_USDHC1_CMD		0x00000021
+			SC_P_USDHC1_DATA0_CONN_USDHC1_DATA0	0x00000021
+			SC_P_USDHC1_DATA1_CONN_USDHC1_DATA1	0x00000021
+			SC_P_USDHC1_DATA2_CONN_USDHC1_DATA2	0x00000021
+			SC_P_USDHC1_DATA3_CONN_USDHC1_DATA3	0x00000021
+			SC_P_USDHC1_VSELECT_CONN_USDHC1_VSELECT	0x00000021
+		>;
+	};
+
+	pinctrl_usdhc2_100mhz: usdhc2grp100mhz {
+		fsl,pins = <
+			SC_P_USDHC1_CLK_CONN_USDHC1_CLK		0x06000041
+			SC_P_USDHC1_CMD_CONN_USDHC1_CMD		0x00000021
+			SC_P_USDHC1_DATA0_CONN_USDHC1_DATA0	0x00000021
+			SC_P_USDHC1_DATA1_CONN_USDHC1_DATA1	0x00000021
+			SC_P_USDHC1_DATA2_CONN_USDHC1_DATA2	0x00000021
+			SC_P_USDHC1_DATA3_CONN_USDHC1_DATA3	0x00000021
+			SC_P_USDHC1_VSELECT_CONN_USDHC1_VSELECT	0x00000021
+		>;
+	};
+
+	pinctrl_usdhc2_200mhz: usdhc2grp200mhz {
+		fsl,pins = <
+			SC_P_USDHC1_CLK_CONN_USDHC1_CLK		0x06000041
+			SC_P_USDHC1_CMD_CONN_USDHC1_CMD		0x00000021
+			SC_P_USDHC1_DATA0_CONN_USDHC1_DATA0	0x00000021
+			SC_P_USDHC1_DATA1_CONN_USDHC1_DATA1	0x00000021
+			SC_P_USDHC1_DATA2_CONN_USDHC1_DATA2	0x00000021
+			SC_P_USDHC1_DATA3_CONN_USDHC1_DATA3	0x00000021
+			SC_P_USDHC1_VSELECT_CONN_USDHC1_VSELECT	0x00000021
+		>;
+	};
+	imx8qxp-mek {
+		pinctrl_i2c0_mipi_lvds0: mipi_lvds0_i2c0_grp {
+			fsl,pins = <
+				SC_P_MIPI_DSI0_I2C0_SCL_MIPI_DSI0_I2C0_SCL	0xc6000020
+				SC_P_MIPI_DSI0_I2C0_SDA_MIPI_DSI0_I2C0_SDA	0xc6000020
+				SC_P_MIPI_DSI0_GPIO0_01_LSIO_GPIO1_IO28		0x00000020
+			>;
+		};
+
+		pinctrl_flexcan1: flexcan0grp {
+			fsl,pins = <
+				SC_P_FLEXCAN0_TX_ADMA_FLEXCAN0_TX		0x00000021
+				SC_P_FLEXCAN0_RX_ADMA_FLEXCAN0_RX		0x00000021
+			>;
+		};
+
+		pinctrl_flexcan3: flexcan3grp {
+			fsl,pins = <
+				SC_P_FLEXCAN2_TX_ADMA_FLEXCAN2_TX		0x00000021
+				SC_P_FLEXCAN2_RX_ADMA_FLEXCAN2_RX		0x00000021
+			>;
+		};
+	};
+};
+
diff --git a/arch/arm64/configs/cl-som-imx8x_defconfig b/arch/arm64/configs/cl-som-imx8x_defconfig
index 19304085c150..9278f1aff92c 100644
--- a/arch/arm64/configs/cl-som-imx8x_defconfig
+++ b/arch/arm64/configs/cl-som-imx8x_defconfig
@@ -154,7 +154,10 @@ CONFIG_VLAN_8021Q_MVRP=y
 CONFIG_LLC2=y
 CONFIG_BPF_JIT=y
 CONFIG_CAN=y
-CONFIG_CAN_FLEXCAN=y
+CONFIG_CAN_RAW=m
+CONFIG_CAN_BCM=m
+CONFIG_CAN_GW=m
+CONFIG_CAN_FLEXCAN=m
 CONFIG_BT=y
 CONFIG_BT_RFCOMM=y
 CONFIG_BT_RFCOMM_TTY=y
-- 
2.11.0

